### PR TITLE
fix(cdc-backfill): ensure snapshot read once for each epoch

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Inspired by https://github.com/risingwavelabs/risingwave/pull/12780

Will pick to main later.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/executor/backfill/cdc/cdc_backfill.rs [hunk 1]
```diff
@@ -232,15 +234,96 @@ impl<S: StateStore> CdcBackfillExecutor<S> {
...
+                                            ?current_pk_pos,
+                                            "force a snapshot read"
+                                        );
+
+                                        let mut snapshot_stream_end = false;
+                                        // If no snapshot read happen, the current_pk_pos is unchanged in previous epoch.
+                                        // We stall the barrier message and force a snapshot read here to avoid starvation of the snapshot stream.
+                                        let (_, snapshot_stream) = backfill_stream.into_inner();
+                                        #[for_await]
+                                        for msg in snapshot_stream {
```
Comment (st1page) [inline src/stream/src/executor/backfill/cdc/cdc_backfill.rs:257]: No need to use `for` here because all path is "break"?

**Review by st1page — APPROVED** [review]
LGTM as a quick patch but when PR to the main branch, I guess we need reuse some common code.

**Comment by kwannoel** [comment]
Why does cdc backfill also encounter this issue? Upstream DB should not have tombstones

**Comment by kwannoel** [comment]
Or is there some other motivation?

**Pull request merged**